install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml

```diff
@@ -276,6 +276,10 @@ spec:
         - mountPath: /sys/fs/bpf
           name: bpf-maps
 {{- end }}
+{{- if not (contains "/run/cilium/cgroupv2" .Values.cgroup.root) }}
+        - mountPath: {{ .Values.cgroup.root }}
```

Comment:
If I'm reading this correctly, `cgroup.root` is used to refer to locations (mountpoints) that, depending on its value, are in the host or in the agent's container. I think that's confusing.

Reply:
(I'm replying to @kkourt 's comment here so that we can have the discussion in a single thread.)

Chatted with Kornilios offline. Looks like there was some confusion around the PR.

If auto-mount is enabled, then cgroup2 fs will be mounted at `cgroupRoot` on both the host and the agent pod. Else,  cgroup2 fs mounted at `cgroupHostRoot` is mounted inside the agent pod at `cgroupRoot`. I can clarify this a bit further.

>one configuration variable that points to the location of the cgroup dir on the host, to be used when autoMount is disabled.

`This is the `cgroupHostRoot` that the PR has introduced. 

>one configuration variable for the cgroup root inside the container, which is what directory that the agent will use.

If this config only points to the directory inside the agent pod, why do we need to expose that as an helm option to users?

Alternatively, we can drop the additional option `cgroupHostRoot`. We hand-waived over some of these details in the datapath meeting.

option 1
- cgroup.autoMount : Enable/disable
- cgroup.root : If `autoMount` is enabled, then agent will mount cgroup2 fs at that path. Else, it'll mount that directory on the host inside the agent pod. In both the cases, we allow users to override the path. However, we'll need to have different defaults in auto-mount enable (`/run/cilium/cgroupv2`) and disable (`/sys/fs/cgroup`) cases. 

option 2
- cgroup.autoMount : Enable/disable
If `autoMount` is enabled, then agent will mount cgroup2 fs at a custom (hard-coded) location `/run/cilium/cgroupv2`.  Users won't be able to override this. 
- cgroup.root : This will be used when users disable auto-mount, and agent will then mount host's `cgroup.root` inside the agent pod. The default will be `/sys/fs/cgroup`.

I think option 2 is simpler.

Reply:
> If this config only points to the directory inside the agent pod, why do we need to expose that as an helm option to users?

:heavy_plus_sign: 

Option 2 sounds simpler and more consistent to me. If `cgroup.autoMount=false`, would we mount `cgroup.root` to `/run/cilium/cgroupv2` inside the container

Reply:
@aditighag, thanks for the clarification! Indeed I misunderstood the PR.

Both options are fine by me. The only comment I would make is that `cgroup.root` seems somewhat similar to the agent option `-cgroup-root` so I would suggest we use a different name to avoid confusing. `cgroup.hostRoot` maybe?

Reply:
> If cgroup.autoMount=false, would we mount cgroup.root to /run/cilium/cgroupv2 inside the container?

This is exactly why I introduced a new option (`cgroupHostRoot`) in the current commits. If we go with option 2, we'll need to have `/run/cilium/cgroupv2` hard-coded in the agent DS. Unless folks have some other ideas?

Reply:
If you don't specify it anywhere, then the default agent value is `""` which it then replaces with `DefaultCgroupRoot = "/var/run/cilium/cgroupv2"`. That seems fine to me.

Reply:
I was referring to the agent daemonset yaml file changes when auto-mount is enabled. 
I'm getting conflicting comments on the PR. 😕 

Where do we want to mount cgroup2 fs inside the cilium agent pod when auto.mount is - 
- enabled
- disabled 

If it's going to be /run/cilium/cgroupv2 in both the cases, then we can have the path hard-coded, and remove the cgroup-root agent configmap option. If we want it to be /run/cilium/cgroupv2 in the 1st case, but same as the host cgroup root for the latter case, then don't we have the same confusion if we rename the current option to cgroupHostRoot? 

IMO, we are overly complicating this. Let's just call it cgroupRoot, and have it better documented.

Reply:
One thing that made it more complicated is it's the same actual directory in the cilium container:

```
# ls -lh /var/run
lrwxrwxrwx 1 root root 4 Apr 16 05:11 /var/run -> /run
```

I think we should stick with `/var/run/cilium/cgroupv2` so we're consistent with the defaults in the daemon, in case for whatever reason these directories diverge at some point in the future.

Then in the DaemonSet, if `cgroup.automount` is enabled, we do the nsenter trick to mount it to this path. If `cgroup.automount` is disabled, then we specify a cgroup hostPath under `volumes` path as the configured argument `cgroup.root` or whatever we call it. Only in the latter case we need a `volumeMount` to mount the path to `/var/run/cilium/cgroupv2` (which will then be automatically used by the daemon defaults without having to specify it anywhere in configmap or DS commandline args).

I'm fine with `cgroup.root` or `cgroup.hostRoot` in the helm config, either way I agree that documenting that it's about the host path is the important piece. I don't think many users care how it looks inside the container.

Reply:
Oh, I think I glossed over some of the detail in the automount side which makes the above comment miss some of the nuance.

For automount I think the goal is to mount at a known Cilium path on the host, then mount that to the consistent internal path. I _think_ if `cgroup.automount=true` it's fine to automount the cgroupfs to either `/var/run/cilium/cgroupv2` or `cgroup.root` on the host, I don't know if makes a big difference.  Then in that case we also need to specify the `volumes` / `volumeMount`, based on wherever we decide the cgroup FS should be automounted. Maybe I'm finally coming around to why this is complicated :sweat_smile:

Reply:
^^ Probably makes sense to have some default for `cgroup.root` and always use that as the host path, either that we try to mount cgroupfs there in the case that `cgroup.automount=true`, or we use that (expected pre-mounted) path to mount into the container in the case that `cgroup.automount=false`. Then we can decide whatever the default for `cgroup.root` should be.